File: graph_mst/prims/src/PrimsDemo.cpp
// C++
// グラフの最小全域木: プリム法 (Prim)

#include <iostream>
#include <vector>
#include <map>
#include <set>
#include <queue>
#include <limits>
#include <algorithm>
#include <tuple>

class GraphData {
private:
    // 隣接リストとしてグラフデータを格納します。
    // キーは頂点、値はその頂点に隣接する頂点とその辺の重みのペアのベクトルです。
    std::map<std::string, std::vector<std::pair<std::string, int>>> _data;

public:
    GraphData() {}

    // グラフの内部データを取得します。
    std::map<std::string, std::vector<std::pair<std::string, int>>>& get() {
        return _data;
    }

    // グラフの全頂点をベクトルとして返します。
    std::vector<std::string> get_vertices() {
        std::vector<std::string> vertices;
        for (const auto& pair : _data) {
            vertices.push_back(pair.first);
        }
        return vertices;
    }

    // グラフの全辺をリストとして返します。
    // 無向グラフの場合、(u, v, weight) の形式で返します。
    std::vector<std::tuple<std::string, std::string, int>> get_edges() {
        std::set<std::tuple<std::string, std::string, int>> edges;
        for (const auto& vertex_pair : _data) {
            for (const auto& neighbor_weight : vertex_pair.second) {
                std::string vertex = vertex_pair.first;
                std::string neighbor = neighbor_weight.first;
                int weight = neighbor_weight.second;
                
                // 辺を正規化してセットに追加 (小さい方の頂点を最初にするなど)
                std::tuple<std::string, std::string, int> edge;
                if (vertex < neighbor) {
                    edge = std::make_tuple(vertex, neighbor, weight);
                } else {
                    edge = std::make_tuple(neighbor, vertex, weight);
                }
                edges.insert(edge);
            }
        }
        return std::vector<std::tuple<std::string, std::string, int>>(edges.begin(), edges.end());
    }

    // 指定された頂点の隣接ノードと辺の重みのリストを返します。
    std::vector<std::pair<std::string, int>> get_neighbors(const std::string& vertex) {
        if (_data.find(vertex) != _data.end()) {
            return _data[vertex];
        }
        return std::vector<std::pair<std::string, int>>(); // 頂点が存在しない場合は空のベクトルを返す
    }

    // 指定された2つの頂点間の辺の重みを返します。
    // 辺が存在しない場合は-1を返します。
    int get_edge_weight(const std::string& vertex1, const std::string& vertex2) {
        if (_data.find(vertex1) != _data.end() && _data.find(vertex2) != _data.end()) {
            for (const auto& neighbor_weight : _data[vertex1]) {
                if (neighbor_weight.first == vertex2) {
                    return neighbor_weight.second;
                }
            }
        }
        return -1; // 辺が存在しない場合
    }

    // 新しい頂点をグラフに追加します。
    bool add_vertex(const std::string& vertex) {
        if (_data.find(vertex) == _data.end()) {
            _data[vertex] = std::vector<std::pair<std::string, int>>();
        }
        return true;
    }

    // 両頂点間に辺を追加します。重みを指定します。
    bool add_edge(const std::string& vertex1, const std::string& vertex2, int weight) {
        if (_data.find(vertex1) == _data.end()) {
            add_vertex(vertex1);
        }
        if (_data.find(vertex2) == _data.end()) {
            add_vertex(vertex2);
        }

        // vertex1 -> vertex2 の辺を追加（重み付き）
        bool edge_exists_v1v2 = false;
        for (auto& neighbor_weight : _data[vertex1]) {
            if (neighbor_weight.first == vertex2) {
                neighbor_weight.second = weight; // 既に存在する場合は重みを更新
                edge_exists_v1v2 = true;
                break;
            }
        }
        if (!edge_exists_v1v2) {
            _data[vertex1].push_back(std::make_pair(vertex2, weight));
        }

        // vertex2 -> vertex1 の辺を追加（重み付き）
        bool edge_exists_v2v1 = false;
        for (auto& neighbor_weight : _data[vertex2]) {
            if (neighbor_weight.first == vertex1) {
                neighbor_weight.second = weight; // 既に存在する場合は重みを更新
                edge_exists_v2v1 = true;
                break;
            }
        }
        if (!edge_exists_v2v1) {
            _data[vertex2].push_back(std::make_pair(vertex1, weight));
        }

        return true;
    }

    // グラフが空かどうかを返します。
    bool is_empty() {
        return _data.empty();
    }

    // グラフを空にします。
    bool clear() {
        _data.clear();
        return true;
    }

    // 最小全域木を計算します
    std::vector<std::tuple<std::string, std::string, int>> get_mst(const std::string* start_vertex = nullptr) {
        std::vector<std::string> vertices = get_vertices();
        if (vertices.empty()) {
            return {}; // グラフが空
        }

        std::string start;
        if (start_vertex == nullptr) {
            start = vertices[0];
        } else if (_data.find(*start_vertex) == _data.end()) {
            std::cout << "ERROR: 開始頂点 " << *start_vertex << " はグラフに存在しません。" << std::endl;
            return {};
        } else {
            start = *start_vertex;
        }

        // MSTに含まれる頂点のセット
        std::set<std::string> in_mst;
        
        // 優先度付きキュー (重み, 現在の頂点, 遷移元の頂点)
        // C++の優先度キューはデフォルトで最大ヒープなので、重みを負にして最小ヒープとして使用
        using PQElement = std::tuple<int, std::string, std::string>;
        std::priority_queue<PQElement, std::vector<PQElement>, std::greater<PQElement>> min_heap;
        
        // MSTを構成する辺のリスト
        std::vector<std::tuple<std::string, std::string, int>> mst_edges;
        
        // 各頂点への最小コスト（MSTに追加する際の辺の重み）
        std::map<std::string, int> min_cost;
        std::map<std::string, std::string> parent;
        
        // 初期化
        for (const auto& v : vertices) {
            min_cost[v] = std::numeric_limits<int>::max();
            parent[v] = "";
        }

        // 開始頂点の処理
        min_cost[start] = 0;
        min_heap.push(std::make_tuple(0, start, "")); // (コスト, 現在の頂点, 遷移元の頂点)

        while (!min_heap.empty()) {
            // 最小コストの辺を持つ頂点を取り出す
            int cost;
            std::string current_vertex, from_vertex;
            std::tie(cost, current_vertex, from_vertex) = min_heap.top();
            min_heap.pop();

            // 既にMSTに含まれている頂点であればスキップ
            if (in_mst.find(current_vertex) != in_mst.end()) {
                continue;
            }

            // 現在の頂点をMSTに追加
            in_mst.insert(current_vertex);

            // MSTに追加された辺を記録 (開始頂点以外)
            if (!from_vertex.empty()) {
                // from_vertex から current_vertex への辺の重みを取得
                int weight = get_edge_weight(from_vertex, current_vertex);
                if (weight != -1) {
                    // 辺を正規化して追加
                    if (from_vertex < current_vertex) {
                        mst_edges.push_back(std::make_tuple(from_vertex, current_vertex, weight));
                    } else {
                        mst_edges.push_back(std::make_tuple(current_vertex, from_vertex, weight));
                    }
                }
            }

            // 現在の頂点に隣接する頂点を調べ、MSTへの追加コストを更新
            auto neighbors_with_weight = get_neighbors(current_vertex);
            for (const auto& neighbor_weight : neighbors_with_weight) {
                std::string neighbor = neighbor_weight.first;
                int weight = neighbor_weight.second;
                
                // 隣接頂点がまだMSTに含まれておらず、現在のコストよりも小さい場合
                if (in_mst.find(neighbor) == in_mst.end() && weight < min_cost[neighbor]) {
                    min_cost[neighbor] = weight;
                    parent[neighbor] = current_vertex;
                    min_heap.push(std::make_tuple(weight, neighbor, current_vertex));
                }
            }
        }

        return mst_edges;
    }
};

int main() {
    std::cout << "Prims TEST -----> start" << std::endl;
    GraphData graph_data;

    graph_data.clear();
    std::vector<std::tuple<std::string, std::string, int>> inputList = {
        {"A", "B", 4}, {"B", "C", 3}, {"B", "D", 2}, {"D", "A", 1}, {"A", "C", 2}, {"B", "D", 2}
    };
    for (const auto& input : inputList) {
        graph_data.add_edge(std::get<0>(input), std::get<1>(input), std::get<2>(input));
    }
    std::cout << "\nグラフの頂点: ";
    for (const auto& vertex : graph_data.get_vertices()) {
        std::cout << vertex << " ";
    }
    std::cout << std::endl;
    
    std::cout << "グラフの辺 (重み付き): ";
    for (const auto& edge : graph_data.get_edges()) {
        std::cout << "(" << std::get<0>(edge) << ", " << std::get<1>(edge) << ", " << std::get<2>(edge) << ") ";
    }
    std::cout << std::endl;
    
    auto outputMst = graph_data.get_mst();
    for (const auto& edge : outputMst) {
        std::cout << "Edge: " << std::get<0>(edge) << " - " << std::get<1>(edge) << ", Weight: " << std::get<2>(edge) << std::endl;
    }
    int total_weight = 0;
    for (const auto& edge : outputMst) {
        total_weight += std::get<2>(edge);
    }
    std::cout << "最小全域木の合計重み: " << total_weight << std::endl;

    graph_data.clear();
    inputList = {
        {"A", "B", 4}, {"C", "D", 4}, {"E", "F", 1}, {"F", "G", 1}
    };
    for (const auto& input : inputList) {
        graph_data.add_edge(std::get<0>(input), std::get<1>(input), std::get<2>(input));
    }
    std::cout << "\nグラフの頂点: ";
    for (const auto& vertex : graph_data.get_vertices()) {
        std::cout << vertex << " ";
    }
    std::cout << std::endl;
    
    std::cout << "グラフの辺 (重み付き): ";
    for (const auto& edge : graph_data.get_edges()) {
        std::cout << "(" << std::get<0>(edge) << ", " << std::get<1>(edge) << ", " << std::get<2>(edge) << ") ";
    }
    std::cout << std::endl;
    
    outputMst = graph_data.get_mst();
    for (const auto& edge : outputMst) {
        std::cout << "Edge: " << std::get<0>(edge) << " - " << std::get<1>(edge) << ", Weight: " << std::get<2>(edge) << std::endl;
    }
    total_weight = 0;
    for (const auto& edge : outputMst) {
        total_weight += std::get<2>(edge);
    }
    std::cout << "最小全域木の合計重み: " << total_weight << std::endl;

    graph_data.clear();
    inputList = {
        {"A", "B", 4}, {"B", "C", 3}, {"D", "E", 5}
    };
    for (const auto& input : inputList) {
        graph_data.add_edge(std::get<0>(input), std::get<1>(input), std::get<2>(input));
    }
    std::cout << "\nグラフの頂点: ";
    for (const auto& vertex : graph_data.get_vertices()) {
        std::cout << vertex << " ";
    }
    std::cout << std::endl;
    
    std::cout << "グラフの辺 (重み付き): ";
    for (const auto& edge : graph_data.get_edges()) {
        std::cout << "(" << std::get<0>(edge) << ", " << std::get<1>(edge) << ", " << std::get<2>(edge) << ") ";
    }
    std::cout << std::endl;
    
    outputMst = graph_data.get_mst();
    for (const auto& edge : outputMst) {
        std::cout << "Edge: " << std::get<0>(edge) << " - " << std::get<1>(edge) << ", Weight: " << std::get<2>(edge) << std::endl;
    }
    total_weight = 0;
    for (const auto& edge : outputMst) {
        total_weight += std::get<2>(edge);
    }
    std::cout << "最小全域木の合計重み: " << total_weight << std::endl;

    graph_data.clear();
    std::cout << "\nグラフの頂点: ";
    for (const auto& vertex : graph_data.get_vertices()) {
        std::cout << vertex << " ";
    }
    std::cout << std::endl;
    
    std::cout << "グラフの辺 (重み付き): ";
    for (const auto& edge : graph_data.get_edges()) {
        std::cout << "(" << std::get<0>(edge) << ", " << std::get<1>(edge) << ", " << std::get<2>(edge) << ") ";
    }
    std::cout << std::endl;
    
    outputMst = graph_data.get_mst();
    for (const auto& edge : outputMst) {
        std::cout << "Edge: " << std::get<0>(edge) << " - " << std::get<1>(edge) << ", Weight: " << std::get<2>(edge) << std::endl;
    }
    total_weight = 0;
    for (const auto& edge : outputMst) {
        total_weight += std::get<2>(edge);
    }
    std::cout << "最小全域木の合計重み: " << total_weight << std::endl;

    std::cout << "\nPrims TEST <----- end" << std::endl;
    return 0;
}

File: graph_mst/prims/src/PrimsDemo.py
# Python
# グラフの最小全域木: プリム法 (Prim)
import heapq

class GraphData:
    def __init__(self):
        # 隣接リストとしてグラフデータを格納します。
        # キーは頂点、値はその頂点に隣接する頂点とその辺の重みのタプルのリストです。
        # 例: { 'A': [('B', 1), ('C', 4)], 'B': [('A', 1), ('C', 2)], ... }
        self._data = {}

    def get(self):
        # グラフの内部データを取得します。
        return self._data

    def get_vertices(self):
        # グラフの全頂点をリストとして返します。
        return list(self._data.keys())
    
    def get_edges(self):
        # グラフの全辺をリストとして返します。
        # 無向グラフの場合、(u, v, weight) の形式で返します。
        # 重複を避けるためにセットを使用します。
        edges = set()
        for vertex in self._data:
            for neighbor, weight in self._data[vertex]:
                # 辺を正規化してセットに追加 (小さい方の頂点を最初にするなど)
                edge = tuple(sorted((vertex, neighbor)))
                edges.add((edge[0], edge[1], weight)) # (u, v, weight) の形式で格納
        return list(edges)
    
    def get_neighbors(self, vertex):
        # 指定された頂点の隣接ノードと辺の重みのリストを返します。
        # 形式: [(隣接頂点, 重み), ...]
        if vertex in self._data:
            return self._data[vertex]
        else:
            return None # 頂点が存在しない場合はNoneを返す

    def get_edge_weight(self, vertex1, vertex2):
        # 指定された2つの頂点間の辺の重みを返します。
        # 辺が存在しない場合はNoneを返します。
        if vertex1 in self._data and vertex2 in self._data:
            for neighbor, weight in self._data[vertex1]:
                if neighbor == vertex2:
                    return weight
        return None # 辺が存在しない場合

    def add_vertex(self, vertex):
        # 新しい頂点をグラフに追加します。
        if vertex not in self._data:
            self._data[vertex] = []
            return True
        # 既に存在する場合は追加しないがTrueを返す（変更なしでも成功とみなす）
        return True
    
    def add_edge(self, vertex1, vertex2, weight):
        # 両頂点間に辺を追加します。重みを指定します。
        # 頂点がグラフに存在しない場合は追加します。
        if vertex1 not in self._data:
            self.add_vertex(vertex1)
        if vertex2 not in self._data:
            self.add_vertex(vertex2)
        
        # 両方向に辺を追加する（無向グラフ）
        # 既に同じ辺が存在する場合は重みを更新するかどうか？
        # 今回は単純に追加する（重複辺はプリム法では問題にならないが、データとしては綺麗でないかも）
        # ここでは、同じ頂点間の辺が存在しない場合のみ追加するように修正します。
        
        # vertex1 -> vertex2 の辺を追加（重み付き）
        edge_exists_v1v2 = False
        for i, (neighbor, _) in enumerate(self._data[vertex1]):
            if neighbor == vertex2:
                self._data[vertex1][i] = (vertex2, weight) # 既に存在する場合は重みを更新
                edge_exists_v1v2 = True
                break
        if not edge_exists_v1v2:
            self._data[vertex1].append((vertex2, weight))

        # vertex2 -> vertex1 の辺を追加（重み付き）
        edge_exists_v2v1 = False
        for i, (neighbor, _) in enumerate(self._data[vertex2]):
            if neighbor == vertex1:
                self._data[vertex2][i] = (vertex1, weight) # 既に存在する場合は重みを更新
                edge_exists_v2v1 = True
                break
        if not edge_exists_v2v1:
            self._data[vertex2].append((vertex1, weight))
        
        return True
    
    def is_empty(self):
        # グラフが空かどうかを返します。
        return len(self._data) == 0
        
    def clear(self):
        # グラフを空にします。
        self._data = {}
        return True

    def get_mst(self, start_vertex=None):
        vertices = self.get_vertices()
        if not vertices:
            return [] # グラフが空

        if start_vertex is None:
            start_vertex = vertices[0]
        elif start_vertex not in self._data:
             print(f"ERROR: 開始頂点 {start_vertex} はグラフに存在しません。")
             return None

        # MSTに含まれる頂点のセット
        in_mst = set()
        # 優先度付きキュー (重み, 現在の頂点, 遷移元の頂点)
        # heapqは最小ヒープなので、重みが小さい辺から取り出される
        min_heap = []
        # MSTを構成する辺のリスト
        mst_edges = []
        # 各頂点への最小コスト（MSTに追加する際の辺の重み）と、その遷移元の頂点を記録
        min_cost = {v: float('inf') for v in vertices}
        parent = {v: None for v in vertices}

        # 開始頂点の処理
        min_cost[start_vertex] = 0
        heapq.heappush(min_heap, (0, start_vertex, None)) # (コスト, 現在の頂点, 遷移元の頂点)

        while min_heap:
            # 最小コストの辺を持つ頂点を取り出す
            cost, current_vertex, from_vertex = heapq.heappop(min_heap)

            # 既にMSTに含まれている頂点であればスキップ
            if current_vertex in in_mst:
                continue

            # 現在の頂点をMSTに追加
            in_mst.add(current_vertex)

            # MSTに追加された辺を記録 (開始頂点以外)
            if from_vertex is not None:
                # from_vertex から current_vertex への辺の重みを取得
                weight = self.get_edge_weight(from_vertex, current_vertex)
                if weight is not None:
                    mst_edges.append(tuple(sorted((from_vertex, current_vertex))) + (weight,)) # 辺を正規化して追加

            # 現在の頂点に隣接する頂点を調べ、MSTへの追加コストを更新
            neighbors_with_weight = self.get_neighbors(current_vertex)
            if neighbors_with_weight: # 隣接する頂点がある場合
                for neighbor, weight in neighbors_with_weight:
                    # 隣接頂点がまだMSTに含まれておらず、現在のコストよりも小さい場合
                    if neighbor not in in_mst and weight < min_cost[neighbor]:
                        min_cost[neighbor] = weight
                        parent[neighbor] = current_vertex
                        heapq.heappush(min_heap, (weight, neighbor, current_vertex))

        # グラフが非連結で、MSTに含まれる頂点が全体の頂点数より少ない場合
        # プリム法は開始頂点を含む連結成分のMSTのみを求めます。
        # この実装では、開始頂点を含む連結成分のMST辺のみを返します。
        # 必要であれば、非連結グラフの森を求めるように拡張することも可能です。

        # 全ての頂点がMSTに含まれたか確認（連結グラフの場合）
        # if len(in_mst) != len(vertices):
        #     print("Warning: グラフは非連結の可能性があります。開始頂点を含む連結成分のMSTを返します。")

        return mst_edges

def main():
    print("Prims TEST -----> start")
    graph_data = GraphData()

    graph_data.clear()
    inputList = [('A', 'B', 4), ('B', 'C', 3), ('B', 'D', 2), ('D', 'A', 1), ('A', 'C', 2), ('B', 'D', 2)]
    for input in inputList:
        graph_data.add_edge(*input)
    print("\nグラフの頂点:", graph_data.get_vertices())
    print("グラフの辺 (重み付き):", graph_data.get_edges())
    outputMst = graph_data.get_mst()
    for edge in outputMst:
        print(f"Edge: {edge[0]} - {edge[1]}, Weight: {edge[2]}")
    total_weight = sum(edge[2] for edge in outputMst)
    print(f"最小全域木の合計重み: {total_weight}")

    graph_data.clear()
    inputList = [('A', 'B', 4), ('C', 'D', 4), ('E', 'F', 1), ('F', 'G', 1)]
    for input in inputList:
        graph_data.add_edge(*input)
    print("\nグラフの頂点:", graph_data.get_vertices())
    print("グラフの辺 (重み付き):", graph_data.get_edges())
    outputMst = graph_data.get_mst()
    for edge in outputMst:
        print(f"Edge: {edge[0]} - {edge[1]}, Weight: {edge[2]}")
    total_weight = sum(edge[2] for edge in outputMst)
    print(f"最小全域木の合計重み: {total_weight}")

    graph_data.clear()
    inputList = [('A', 'B', 4), ('B', 'C', 3), ('D', 'E', 5)]
    for input in inputList:
        graph_data.add_edge(*input)
    print("\nグラフの頂点:", graph_data.get_vertices())
    print("グラフの辺 (重み付き):", graph_data.get_edges())
    outputMst = graph_data.get_mst()
    for edge in outputMst:
        print(f"Edge: {edge[0]} - {edge[1]}, Weight: {edge[2]}")
    total_weight = sum(edge[2] for edge in outputMst)
    print(f"最小全域木の合計重み: {total_weight}")

    graph_data.clear()
    inputList = []
    for input in inputList:
        graph_data.add_edge(*input)
    print("\nグラフの頂点:", graph_data.get_vertices())
    print("グラフの辺 (重み付き):", graph_data.get_edges())
    outputMst = graph_data.get_mst()
    for edge in outputMst:
        print(f"Edge: {edge[0]} - {edge[1]}, Weight: {edge[2]}")
    total_weight = sum(edge[2] for edge in outputMst)
    print(f"最小全域木の合計重み: {total_weight}")

    print("\nPrims TEST <----- end")

if __name__ == "__main__":
    main()